### PR TITLE
Add support for multiple holochain versions

### PR DESCRIPTION
**Pull request by matthme**
Adds support for running multiple holochain versions, including custom binaries and externally running holochain versions.

### CODE REVIEW

File: src/main/filesystem.ts [hunk 1]
```diff
@@ -26,69 +28,88 @@ class LauncherFileSystem {
...
+      switch (semver.minor(version)) {
+        case 0:
+          return `0.0.${semver.patch(version)}`;
+        default:
+          return `0.${semver.minor(version)}.x`;
+      }
+    default:
+      return `${semver.major(version)}.x.x`;
+  }
+}
```
Comment (mrruby) [inline src/main/filesystem.ts]: how about?

```
function breakingVersion(version) {
  if (!semver.valid(version)) {
    throw new Error('Version is not valid semver.');
  }

  if (semver.prerelease(version)) {
    return version;
  }

  const major = semver.major(version);
  const minor = semver.minor(version);
  const patch = semver.patch(version);

  if (major === 0) {
    return minor === 0 ? `0.0.${patch}` : `0.${minor}.x`;
  }

  return `${major}.x.x`;
}
```

File: src/main/holochainManager.ts [hunk 2]
```diff
@@ -148,16 +220,34 @@ export class HolochainManager {
...
+      switch (semver.minor(version)) {
+        case 0:
+          return `0.0.${semver.patch(version)}`;
+        default:
+          return `0.${semver.minor(version)}.x`;
+      }
+    default:
+      return `${semver.major(version)}.x.x`;
+  }
+}
```
Comment (mrruby) [inline src/main/holochainManager.ts]: we can make a helper function from my comment abour and reuse it here

File: src/main/logs.ts [hunk 3]
```diff
@@ -107,3 +107,17 @@ function createLairLogger(
...
+    identifier = `HOLOCHAIN ${holochainData.version.version} @ partition ${holochainData.partition}`;
+  } else if (holochainData.version.type === 'custom-path') {
+    identifier = `HOLOCHAIN CUSTOM BINARY @ partition ${holochainData.partition}`;
+  } else if (holochainData.version.type === 'running-external') {
+    identifier = `HOLOCHAIN EXTERNAL BINARY @ partition ${holochainData.partition}`;
+  } else {
+    identifier = `HOLOCHAIN unknown`;
+  }
+  return identifier;
+}
```
Comment (mrruby) [inline src/main/logs.ts]: Let's use early reutnr pattern insted of reassigning varable

```
function identifierFromHolochainData(holochainData) {
  if (holochainData.version.type === 'built-in') {
    return `HOLOCHAIN ${holochainData.version.version} @ partition ${holochainData.partition}`;
  }
  
  if (holochainData.version.type === 'custom-path') {
    return `HOLOCHAIN CUSTOM BINARY @ partition ${holochainData.partition}`;
  }
  
  if (holochainData.version.type === 'running-external') {
    return `HOLOCHAIN EXTERNAL BINARY @ partition ${holochainData.partition}`;
  }
  
  return `HOLOCHAIN unknown`;
}
```

File: src/types/holochain.ts [hunk 4]
```diff
@@ -1,9 +1,38 @@
...
+      // Externally running holochain binary. E.g. started via a terminal.
+      type: 'running-external';
+    };
+
+export interface RunningHolochain {
+  version: HolochainVersion;
+  partition: string;
+  appPort: number;
+}
+
```
Comment (mrruby) [inline src/types/holochain.ts]: can we use types insted of interfaces?

**Review by mrruby — CHANGES_REQUESTED** [review]

File: src/types/holochain.ts (hunk 4 above)
Comment (matthme) [reply src/types/holochain.ts]: Yes. I pushed a new commit, can you recheck?

**Review by mrruby — APPROVED** [review]

**Pull request merged**